ENH: Custom Parser Directories Redesign

joeykleingers opened the pull request:
+ Custom parser directories list has been converted into a table that displays the following information for each row:
   + Custom parser directory path
   + Names of all parsers found in the given custom parser directory path
   + Checkbox to enable/disable entire parser directories
   + Gear icon that opens the given parser directory's config YAML file using the default OS application, allowing the user to easily edit it to add/remove parsers.

_Note:_ The default parser directory has a star icon next to the path and it is listed in the first row of the table.  It cannot be removed from the table.

+ It is now possible to update a custom parser directory's config YAML file outside the MetaForge application.. MetaForge will detect the update the update the list of loaded parsers accordingly.
+ Automatically add a skeleton yaml file to a custom parser directory if it does not have one already.

joeykleingers committed:
Redesign custom parser list in Preferences dialog.

The custom parser list has been converted into a table and now displays
each custom parser directory path with the names of all the parsers
found within each directory.  Custom parser directories can now be enabled
and disabled using the checkbox.  Additionally, each parser directory's
config file can be edited by double clicking on the gear icon in a given
row of the table.

The default parser directory is now included as the first item in the
table and cannot be removed from the table.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Sort the parsers combo box.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Add some detailed instructions to generated skeleton YAML config files.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Fix accidental variable name collision and unwind parser match line.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

cryos reviewed (approved):
Some great additional features, really like the display of the parser file path. I noted a non-obvious failure when a parser can't be loaded, we talked over Slack about it. It would be great to make that failure more obvious and avoid crashing, but it was due to me not having a Python module in my environment. Also miss the default empty parser on first load where it chooses the right one with a best guess.

I think this can be merged as-is, with some follow up, or you can address the minor things I pointed out before merging. Happy to test further.

joeykleingers committed:
Update parser table to list parsers instead of directories.
Parser table now displays more information about each parser.
Parsers that cannot be loaded display an error message in the table.
MetaForge now reloads parsers whose underlying file is updated on the file system.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Remove missing constant reference.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Set a red background for parsers that have load errors.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Implement a search bar for the parsers table, and enable sorting by column.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers commented:
In MetaForge, we have two different locations where we can view the currently loaded parsers.  The first location is in the preferences panel where a table lists all the directories the user has loaded, as well as the parsers in each directory that were loaded (new to this PR).  The second location is a combo box on the Create Template side of the application, which simply lists all the available parsers alphabetically and the user can pick which one to use.

The model that backs up the preferences panel table is a multi-level model because each table row displays a parser directory which can contain many parsers within it, whereas the model that backs up the combo box is a linear model that simply displays all the currently loaded parsers from those directories.  I was originally keeping these as two separate models because one is linear and one isn't, but this design was causing minor bugs to appear, simply due to all the bookkeeping that had to happen behind the scenes to keep both models consistent when parser directories were added/removed from the preferences panel table.

I went down the road of attempting to write the adapter code necessary to have the multi-level model internally use the combo box's linear model, but this is not straight forward, especially if we want to use the Qt modeling system to automatically propagate model changes up to both views (the cleanest way of using Qt model/view architecture).

To fix the double model bugs and achieve parser model consistency throughout the application, I decided to just have the preferences panel table use the same linear model that the combo box is using, and update the preferences panel table UI to allow the user to directly pick parser files instead of directories.  Now both views display that same linear parser model, and all those bugs have disappeared.

In an effort to keep the table easy-to-use, I have also implemented the following:

1. Parser multi-selection in the table to be able to unload multiple parsers at once.
2. File multi-selection so that users can load multiple parser files at once.
3. A search bar for filtering, and the ability to sort columns, so that users can easily find specific parsers in the table.

This solution also allows us to discard our custom parser yaml files as well, since we are loading parser files directly into the table instead of loading directories.

If, in the future, we decide that we really do still want to load directories of parsers instead of the parsers themselves, there's probably a way to make that happen, it's just a non-trivial problem that will require some time to address.
